src/combinatorics/simple_graph/connectivity.lean

```diff
@@ -208,6 +215,26 @@ lemma append_assoc : Π {u v w x : V} (p : G.walk u v) (q : G.walk v w) (r : G.w
   (hu : u = u') (hv : v = v') (hw : w = w') :
   (p.copy hu hv).append (q.copy hv hw) = (p.append q).copy hu hw := by { subst_vars, refl }
 
+@[simp] lemma concat_nil {u v : V} (h : G.adj u v) : nil.concat h = cons h nil := rfl
+
+@[simp] lemma concat_cons {u v w x : V} (h : G.adj u v) (p : G.walk v w) (h' : G.adj w x) :
+  (cons h p).concat h' = cons h (p.concat h') := rfl
+
+lemma append_concat {u v w x : V} (p : G.walk u v) (q : G.walk v w) (h : G.adj w x) :
+  p.append (q.concat h) = (p.append q).concat h := append_assoc _ _ _
+
+lemma concat_append {u v w x : V} (p : G.walk u v) (h : G.adj v w) (q : G.walk w x) :
+  (p.concat h).append q = p.append (cons h q) :=
+by rw [concat_eq_append, ← append_assoc, cons_nil_append]
+
+/-- A non-trivial `cons` walk is representable as a `concat` walk. -/
+def cons_eq_concat : Π {u v w : V} (h : G.adj u v) (p : G.walk v w),
+  Σ' (x : V) (q : G.walk u x) (h' : G.adj x w), cons h p = q.concat h'
```

Comment:
Are you using `psigma` to preserve computability? In similar situations, we either used `Exists` or a definition along with theorems. This is a weird mix of both.

Reply:
It's not for computability per se; it's not an `Exists` because there is a unique `x`, `q`, and `h'` with that property.

I didn't think very hard about this function, and it's more or less a translation of a function from Floris's HOTT library in Lean 2. I'd sooner delete it than really have to rework it, though maybe it wouldn't be so bad to define an equivalence (one defect with this function is how one-directional it is).

Reply:
I went and made an equivalence, but then ended up deleting it because in the end it didn't seem like it'd be that useful. This definition is now a lemma proving `Exists`.